project/src/backend/sdl3/SDLSystem.cpp

```diff
@@ -0,0 +1,1041 @@
+
+			mode.height = displayMode->h;
+
+			switch (displayMode->format) {
+
+				case SDL_PIXELFORMAT_ARGB8888:
+
+					mode.pixelFormat = ARGB32;
+					break;
+
+				case SDL_PIXELFORMAT_BGRA8888:
+				case SDL_PIXELFORMAT_BGRX8888:
+
+					mode.pixelFormat = BGRA32;
+					break;
+
+				default:
+
+					mode.pixelFormat = RGBA32;
+
+			}
+
+			mode.refreshRate = displayMode->refresh_rate;
+			mode.width = displayMode->w;
+
+			alloc_field (display, id_currentMode, (value)mode.Value ());
+
+			int numDisplayModes;
+			SDL_DisplayMode **displayModes = SDL_GetFullscreenDisplayModes (id, &numDisplayModes);
+			value supportedModes = alloc_array (numDisplayModes);
+
+			for (int i = 0; i < numDisplayModes; i++) {
+
+				SDL_DisplayMode *sdlDisplayMode = displayModes[i];
+				mode.height = sdlDisplayMode->h;
+
+				switch (sdlDisplayMode->format) {
+
+					case SDL_PIXELFORMAT_ARGB8888:
+
+						mode.pixelFormat = ARGB32;
+						break;
+
+					case SDL_PIXELFORMAT_BGRA8888:
+					case SDL_PIXELFORMAT_BGRX8888:
+
+						mode.pixelFormat = BGRA32;
+						break;
+
+					default:
+
+						mode.pixelFormat = RGBA32;
+
+				}
+
+				mode.refreshRate = sdlDisplayMode->refresh_rate;
+				mode.width = sdlDisplayMode->w;
+
+				val_array_set_i (supportedModes, i, (value)mode.Value ());
+
+			}
+
+			alloc_field (display, id_supportedModes, supportedModes);
+			return display;
+
+		} else {
+
+			const int id_bounds = hl_hash_utf8 ("bounds");
+			const int id_currentMode = hl_hash_utf8 ("currentMode");
+			const int id_dpi = hl_hash_utf8 ("dpi");
+			const int id_height = hl_hash_utf8 ("height");
+			const int id_name = hl_hash_utf8 ("name");
+			const int id_orientation = hl_hash_utf8 ("orientation");
+			const int id_pixelFormat = hl_hash_utf8 ("pixelFormat");
+			const int id_refreshRate = hl_hash_utf8 ("refreshRate");
+			const int id_supportedModes = hl_hash_utf8 ("supportedModes");
+			const int id_width = hl_hash_utf8 ("width");
+			const int id_safeArea = hl_hash_utf8 ("safeArea");
+			const int id_x = hl_hash_utf8 ("x");
+			const int id_y = hl_hash_utf8 ("y");
+
+			if (id == 0) {
+
+				return 0;
+
+			}
+
+			const char* displayName = SDL_GetDisplayName (id);
+			if (displayName == NULL) {
+
+				return 0;
+
+			}
+
+			vdynamic* display = (vdynamic*)hl_alloc_dynobj ();
+
+			char* _displayName = (char*)malloc(strlen(displayName) + 1);
+			strcpy (_displayName, displayName);
+			hl_dyn_setp (display, id_name, &hlt_bytes, _displayName);
+
+			SDL_Rect bounds = { 0, 0, 0, 0 };
+			SDL_GetDisplayBounds (id, &bounds);
+
+			vdynamic* _bounds = (vdynamic*)hl_alloc_dynobj ();
+			hl_dyn_seti (_bounds, id_x, &hlt_i32, bounds.x);
+			hl_dyn_seti (_bounds, id_y, &hlt_i32, bounds.y);
+			hl_dyn_seti (_bounds, id_width, &hlt_i32, bounds.w);
+			hl_dyn_seti (_bounds, id_height, &hlt_i32, bounds.h);
+
+			hl_dyn_setp (display, id_bounds, &hlt_dynobj, _bounds);
+
+			Rectangle safeAreaInsets;
+			Display::GetSafeAreaInsets(id, &safeAreaInsets);
```

Comment:
Was getting a crash on macOS when targeting Hashlink, this fixes it:

```suggestion
			Display::GetSafeAreaInsets(id - 1, &safeAreaInsets);
```